### PR TITLE
Cleanup SMPP Message ID correlation for messages waiting for DLR

### PR DESCRIPTION
**Pull request by hrosa**
**What this PR does / why we need it**:

Enforces uniqueness in such way that no more than one SMS Message becomes correlated to the same SMPP Message ID.

**Which issue(s) this PR fixes**:
Fixes BS-232

**Special notes for your reviewer**:
none

### CODE REVIEW

File: restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/SmsMessageFilter.java [hunk 1]
```diff
@@ -19,38 +19,48 @@
...
-    private final Integer limit;
-    private final Integer offset;
-    private final String instanceid;
+    private static final DateFormat DATE_FORMAT = new SimpleDateFormat("yyyy-MM-dd");
+
+    private String accountSid;
+    private List<String> accountSidSet; // if not-null we need the cdrs that belong to several accounts
+    private String recipient;
+    private String sender;
+    private String startTime;  // to initialize it pass string arguments with  yyyy-MM-dd format
```
Comment (maria-farooq) [inline restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/SmsMessageFilter.java]: what was the reason to change the type from Date to String

File: restcomm/restcomm.dao/src/main/resources/org/restcomm/connect/dao/mybatis/sms-messages.xml [hunk 2]
```diff
@@ -117,8 +117,12 @@
                 AND "date_created" &gt;= #{startTime}
             </if>
             <if test="endTime != null">
-                    AND "date_created" &lt;= DATE_ADD(#{endTime},INTERVAL 1 DAY)
+                AND "date_created" &lt;= DATE_ADD(#{endTime},INTERVAL 1 DAY)
+            </if>
+            <if test="smppMessageId != null">
```
Comment (maria-farooq) [inline restcomm/restcomm.dao/src/main/resources/org/restcomm/connect/dao/mybatis/sms-messages.xml]: we need to make this change for `sms-messages.xml` files in `restcomm.application` module as they are the ones that are packaged and used on production

File: restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/SmsMessageFilter.java (hunk 1 above)
Comment (hrosa) [reply restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/SmsMessageFilter.java]: As far as the filter is concerned, the date is a String with some date-like pattern.
If you check the builder, it accepts a Date which is internally parsed to the desired format. IMO much cleaner and avoids parsing exception when building these objects.

File: restcomm/restcomm.dao/src/main/resources/org/restcomm/connect/dao/mybatis/sms-messages.xml (hunk 2 above)
Comment (hrosa) [reply restcomm/restcomm.dao/src/main/resources/org/restcomm/connect/dao/mybatis/sms-messages.xml]: good point, agree

File: restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/SmsMessagesDao.java [hunk 3]
```diff
@@ -53,4 +54,6 @@
     // Support for filtering of message list result, Issue 1395
     Integer getTotalSmsMessage(SmsMessageFilter filter);
     List<SmsMessage> getSmsMessages(SmsMessageFilter filter);
+
+    List<SmsMessage> findBySmppMessageIdAndDateCreatedGreaterOrEqualThanOrderedByDateCreatedDesc(String smppMessageId, DateTime startDate);
```
Comment (jaimecasero) [inline restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/SmsMessagesDao.java]: i would prefer a Filter based API approach as in https://github.com/RestComm/Restcomm-Connect/blob/master/restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/IncomingPhoneNumbersDao.java#L48

youll see it supports sorting as well https://github.com/RestComm/Restcomm-Connect/blob/master/restcomm/restcomm.dao/src/main/java/org/restcomm/connect/dao/entities/IncomingPhoneNumberFilter.java#L34

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java [hunk 4]
```diff
@@ -143,15 +141,32 @@ public void onReceive(Object message) throws Exception {
...
+            final DLRPayload deliveryReceipt = (DLRPayload) message;
+            final String smppMessageId = deliveryReceipt.getId();
+            final SmsMessage.Status deliveryStatus = deliveryReceipt.getStat();
+
+            if (logger.isDebugEnabled()) {
+                logger.debug("DLR Received for SMPP Message " + deliveryReceipt.getId() + " with status " + deliveryStatus);
+            }
+
+            // Find all messages correlated with SMPP Message ID in last three days
+            final List<SmsMessage> smsMessages = this.storage.getSmsMessagesDao().findBySmppMessageIdAndDateCreatedGreaterOrEqualThanOrderedByDateCreatedDesc(smppMessageId, DateTime.now().minusDays(3));
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java]: i thought we agreed this query was actually for Submit part, not DLR

File: restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java [hunk 5]
```diff
@@ -143,15 +141,32 @@ public void onReceive(Object message) throws Exception {
...
+            // Update status of messages and remove correlation with SMPP Message ID
+            // IMPORTANT: First message in the results list is considered the real target of the DLR
+            // BS-232: Other "pending" messages that might share correlation with same SMPP Message ID will have their status forcefully updated to SENT
+            final int smsCount = smsMessages.size();
+            if (smsCount == 0) {
+                logger.warning("responseMessageId=" + deliveryReceipt.getId() + " was never received! ");
             } else {
-                // Clean correlation to SMPP Message ID because SMPP identifiers are may repeat after a given time frame
-                smsMessagesDao.updateSmsMessage(smsMessage.setSmppMessageId(null).setStatus(dLRPayload.getStat()));
+                for (int index = 0; index < smsCount; index++) {
```
Comment (jaimecasero) [inline restcomm/restcomm.sms/src/main/java/org/restcomm/connect/sms/smpp/SmppMessageHandler.java]: same for this consistency check, i thought it was for submit Resp

**Pull request merged**